<?xml version="1.0" encoding="utf-8"?>
<?xml-stylesheet type="text/xsl" href="../assets/xml/rss.xsl" media="all"?><rss version="2.0" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:atom="http://www.w3.org/2005/Atom"><channel><title>The Cloistered Monkey (Posts about troubleshooting)</title><link>https://necromuralist.github.io/</link><description></description><atom:link href="https://necromuralist.github.io/categories/troubleshooting.xml" rel="self" type="application/rss+xml"></atom:link><language>en</language><copyright>Contents © 2024 &lt;a href="mailto:cloisteredmonkey.jmark@slmail.me"&gt;Cloistered Monkey&lt;/a&gt; 
&lt;div id="license"xmlns:cc="http://creativecommons.org/ns#" &gt;This work is licensed under
&lt;a href="http://creativecommons.org/licenses/by/4.0/?ref=chooser-v1" target="_blank" rel="license noopener noreferrer" style="display:inline-block;"&gt;CC BY 4.0
&lt;img src="https://mirrors.creativecommons.org/presskit/icons/cc.svg?ref=chooser-v1"&gt;
&lt;img src="https://mirrors.creativecommons.org/presskit/icons/by.svg?ref=chooser-v1"&gt;&lt;/a&gt;
&lt;/div&gt;
</copyright><lastBuildDate>Tue, 28 May 2024 00:09:15 GMT</lastBuildDate><generator>Nikola (getnikola.com)</generator><docs>http://blogs.law.harvard.edu/tech/rss</docs><item><title>Ubuntu 22.04, Python 3.11 and the "emacs-jupyter Symbol's variable is void" Error</title><link>https://necromuralist.github.io/posts/python-311-and-emacs-jupyter-symbols-variable-is-void-error/</link><dc:creator>Cloistered Monkey</dc:creator><description>&lt;div id="outline-container-orgf87be14" class="outline-2"&gt;
&lt;h2 id="orgf87be14"&gt;What This Is About&lt;/h2&gt;
&lt;div class="outline-text-2" id="text-orgf87be14"&gt;
&lt;p&gt;
Ubuntu 22.04 no longer let's you install python packages globally using pip by default (you can override it but they warn you not to). This has caused a cascade of broken parts on my system, since I use python so much. This particular case started with me trying to start the &lt;a href="https://docs.jupyter.org/en/latest/projects/kernels.html"&gt;&lt;code&gt;jupyter kernel&lt;/code&gt;&lt;/a&gt; so that I could run some python code in org-mode and getting (what looked like) an error and fixing it ended up uncovering the fact that working with the new policy for pip broke my emacs setup a little too, so this is a dump of how I got it back up and running again. I recorded it as I was fixing things so there might be a better way, but this is the first pass I took.
&lt;/p&gt;
&lt;/div&gt;
&lt;/div&gt;

&lt;div id="outline-container-org1c0defd" class="outline-2"&gt;
&lt;h2 id="org1c0defd"&gt;The Jupyter Kernel Warning&lt;/h2&gt;
&lt;div class="outline-text-2" id="text-org1c0defd"&gt;
&lt;p&gt;
This is what happened when I tried to start the jupyter kernel.
&lt;/p&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;&lt;span class="o"&gt;(&lt;/span&gt;Ape-Iron&lt;span class="o"&gt;)&lt;/span&gt;&lt;span class="w"&gt; &lt;/span&gt;hades@erebus&lt;span class="w"&gt; &lt;/span&gt;~&amp;gt;&lt;span class="w"&gt; &lt;/span&gt;jupyter&lt;span class="w"&gt; &lt;/span&gt;kernel
&lt;span class="o"&gt;[&lt;/span&gt;KernelApp&lt;span class="o"&gt;]&lt;/span&gt;&lt;span class="w"&gt; &lt;/span&gt;Starting&lt;span class="w"&gt; &lt;/span&gt;kernel&lt;span class="w"&gt; &lt;/span&gt;&lt;span class="s1"&gt;'python3'&lt;/span&gt;
&lt;span class="m"&gt;0&lt;/span&gt;.00s&lt;span class="w"&gt; &lt;/span&gt;-&lt;span class="w"&gt; &lt;/span&gt;Debugger&lt;span class="w"&gt; &lt;/span&gt;warning:&lt;span class="w"&gt; &lt;/span&gt;It&lt;span class="w"&gt; &lt;/span&gt;seems&lt;span class="w"&gt; &lt;/span&gt;that&lt;span class="w"&gt; &lt;/span&gt;frozen&lt;span class="w"&gt; &lt;/span&gt;modules&lt;span class="w"&gt; &lt;/span&gt;are&lt;span class="w"&gt; &lt;/span&gt;being&lt;span class="w"&gt; &lt;/span&gt;used,&lt;span class="w"&gt; &lt;/span&gt;which&lt;span class="w"&gt; &lt;/span&gt;may
&lt;span class="m"&gt;0&lt;/span&gt;.00s&lt;span class="w"&gt; &lt;/span&gt;-&lt;span class="w"&gt; &lt;/span&gt;make&lt;span class="w"&gt; &lt;/span&gt;the&lt;span class="w"&gt; &lt;/span&gt;debugger&lt;span class="w"&gt; &lt;/span&gt;miss&lt;span class="w"&gt; &lt;/span&gt;breakpoints.&lt;span class="w"&gt; &lt;/span&gt;Please&lt;span class="w"&gt; &lt;/span&gt;pass&lt;span class="w"&gt; &lt;/span&gt;-Xfrozen_modules&lt;span class="o"&gt;=&lt;/span&gt;off
&lt;span class="m"&gt;0&lt;/span&gt;.00s&lt;span class="w"&gt; &lt;/span&gt;-&lt;span class="w"&gt; &lt;/span&gt;to&lt;span class="w"&gt; &lt;/span&gt;python&lt;span class="w"&gt; &lt;/span&gt;to&lt;span class="w"&gt; &lt;/span&gt;disable&lt;span class="w"&gt; &lt;/span&gt;frozen&lt;span class="w"&gt; &lt;/span&gt;modules.
&lt;span class="m"&gt;0&lt;/span&gt;.00s&lt;span class="w"&gt; &lt;/span&gt;-&lt;span class="w"&gt; &lt;/span&gt;Note:&lt;span class="w"&gt; &lt;/span&gt;Debugging&lt;span class="w"&gt; &lt;/span&gt;will&lt;span class="w"&gt; &lt;/span&gt;proceed.&lt;span class="w"&gt; &lt;/span&gt;Set&lt;span class="w"&gt; &lt;/span&gt;&lt;span class="nv"&gt;PYDEVD_DISABLE_FILE_VALIDATION&lt;/span&gt;&lt;span class="o"&gt;=&lt;/span&gt;&lt;span class="m"&gt;1&lt;/span&gt;&lt;span class="w"&gt; &lt;/span&gt;to&lt;span class="w"&gt; &lt;/span&gt;disable&lt;span class="w"&gt; &lt;/span&gt;this&lt;span class="w"&gt; &lt;/span&gt;val
idation.
&lt;span class="o"&gt;[&lt;/span&gt;KernelApp&lt;span class="o"&gt;]&lt;/span&gt;&lt;span class="w"&gt; &lt;/span&gt;Connection&lt;span class="w"&gt; &lt;/span&gt;file:&lt;span class="w"&gt; &lt;/span&gt;/home/hades/.local/share/jupyter/runtime/kernel-a57a8231-bfea-468
&lt;span class="m"&gt;0&lt;/span&gt;-9f8b-6bf1b1e3a7ac.json
&lt;span class="o"&gt;[&lt;/span&gt;KernelApp&lt;span class="o"&gt;]&lt;/span&gt;&lt;span class="w"&gt; &lt;/span&gt;To&lt;span class="w"&gt; &lt;/span&gt;connect&lt;span class="w"&gt; &lt;/span&gt;a&lt;span class="w"&gt; &lt;/span&gt;client:&lt;span class="w"&gt; &lt;/span&gt;--existing&lt;span class="w"&gt; &lt;/span&gt;kernel-a57a8231-bfea-4680-9f8b-6bf1b1e3a7ac.json
&lt;span class="m"&gt;0&lt;/span&gt;.00s&lt;span class="w"&gt; &lt;/span&gt;-&lt;span class="w"&gt; &lt;/span&gt;Debugger&lt;span class="w"&gt; &lt;/span&gt;warning:&lt;span class="w"&gt; &lt;/span&gt;It&lt;span class="w"&gt; &lt;/span&gt;seems&lt;span class="w"&gt; &lt;/span&gt;that&lt;span class="w"&gt; &lt;/span&gt;frozen&lt;span class="w"&gt; &lt;/span&gt;modules&lt;span class="w"&gt; &lt;/span&gt;are&lt;span class="w"&gt; &lt;/span&gt;being&lt;span class="w"&gt; &lt;/span&gt;used,&lt;span class="w"&gt; &lt;/span&gt;which&lt;span class="w"&gt; &lt;/span&gt;may
&lt;span class="m"&gt;0&lt;/span&gt;.00s&lt;span class="w"&gt; &lt;/span&gt;-&lt;span class="w"&gt; &lt;/span&gt;make&lt;span class="w"&gt; &lt;/span&gt;the&lt;span class="w"&gt; &lt;/span&gt;debugger&lt;span class="w"&gt; &lt;/span&gt;miss&lt;span class="w"&gt; &lt;/span&gt;breakpoints.&lt;span class="w"&gt; &lt;/span&gt;Please&lt;span class="w"&gt; &lt;/span&gt;pass&lt;span class="w"&gt; &lt;/span&gt;-Xfrozen_modules&lt;span class="o"&gt;=&lt;/span&gt;off
&lt;span class="m"&gt;0&lt;/span&gt;.00s&lt;span class="w"&gt; &lt;/span&gt;-&lt;span class="w"&gt; &lt;/span&gt;to&lt;span class="w"&gt; &lt;/span&gt;python&lt;span class="w"&gt; &lt;/span&gt;to&lt;span class="w"&gt; &lt;/span&gt;disable&lt;span class="w"&gt; &lt;/span&gt;frozen&lt;span class="w"&gt; &lt;/span&gt;modules.
&lt;span class="m"&gt;0&lt;/span&gt;.00s&lt;span class="w"&gt; &lt;/span&gt;-&lt;span class="w"&gt; &lt;/span&gt;Note:&lt;span class="w"&gt; &lt;/span&gt;Debugging&lt;span class="w"&gt; &lt;/span&gt;will&lt;span class="w"&gt; &lt;/span&gt;proceed.&lt;span class="w"&gt; &lt;/span&gt;Set&lt;span class="w"&gt; &lt;/span&gt;&lt;span class="nv"&gt;PYDEVD_DISABLE_FILE_VALIDATION&lt;/span&gt;&lt;span class="o"&gt;=&lt;/span&gt;&lt;span class="m"&gt;1&lt;/span&gt;&lt;span class="w"&gt; &lt;/span&gt;to&lt;span class="w"&gt; &lt;/span&gt;disable&lt;span class="w"&gt; &lt;/span&gt;this&lt;span class="w"&gt; &lt;/span&gt;validation.
&lt;/pre&gt;&lt;/div&gt;

&lt;p&gt;
According to &lt;a href="https://stackoverflow.com/questions/75114841/debugger-warning-from-ipython-frozen-modules"&gt;this Stack Overflow post&lt;/a&gt; the output, though scary-looking, is only a warning, and you should be able to ignore it. It's happening because python 3.11 uses a "frozen" version of python with code objects for some of the built-in python modules that get loaded when the interpreter starts up already pre-allocated in order to reduce their load time during python's start up (i.e. they set it up to start faster), and their doing this means that the debugger might not work correctly - but since I'm not using the debugger, it shouldn't matter.
&lt;/p&gt;

&lt;ul class="org-ul"&gt;
&lt;li&gt;&lt;a href="https://docs.python.org/3/whatsnew/3.11.html#faster-startup"&gt;Python Note on their freezing the modules&lt;/a&gt;&lt;/li&gt;
&lt;li&gt;&lt;a href="https://wiki.python.org/moin/Freeze"&gt;Python documentation of "freezing"&lt;/a&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;p&gt;
Ah, but there's always a problem lurking behind the advice to ignore "harmless warnings". Even with the kernel running, I couldn't get python/jupyter to work in my org-babel source blocks, so there was more to do.
&lt;/p&gt;
&lt;/div&gt;
&lt;/div&gt;

&lt;div id="outline-container-org962b1db" class="outline-2"&gt;
&lt;h2 id="org962b1db"&gt;Getting emacs-jupyter Working&lt;/h2&gt;
&lt;div class="outline-text-2" id="text-org962b1db"&gt;
&lt;/div&gt;
&lt;div id="outline-container-org4c60c2f" class="outline-3"&gt;
&lt;h3 id="org4c60c2f"&gt;The Problem&lt;/h3&gt;
&lt;div class="outline-text-3" id="text-org4c60c2f"&gt;
&lt;p&gt;
The first clue as to what might be happing was this line in emacs' startup messages.
&lt;/p&gt;

&lt;pre class="example" id="org16b2cb1"&gt;
Symbol’s function definition is void: org-babel-execute:jupyter-python
&lt;/pre&gt;

&lt;p&gt;
It looked like &lt;a href="https://github.com/emacs-jupyter/jupyter"&gt;emacs-jupyter&lt;/a&gt; wasn't loading properly. There was also this message in the output:
&lt;/p&gt;

&lt;pre class="example" id="orga915c96"&gt;
Error retrieving kernelspecs: (json-number-format 5)
&lt;/pre&gt;

&lt;p&gt;
Searching for that error-message brought up this bug-report on github:
&lt;/p&gt;

&lt;ul class="org-ul"&gt;
&lt;li&gt;&lt;a href="https://github.com/emacs-jupyter/jupyter/issues/436"&gt;https://github.com/emacs-jupyter/jupyter/issues/436&lt;/a&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;p&gt;
Wherein the author of the bug-report mentions that loading emacs-jupyter is failing because it's trying to parse the output of jupyter and the warnings I was seeing causes it to fail (the bug-report references a different jupyter command, but the problematic output is the same).
&lt;/p&gt;
&lt;/div&gt;
&lt;/div&gt;

&lt;div id="outline-container-org0113c1a" class="outline-3"&gt;
&lt;h3 id="org0113c1a"&gt;Testing Turning Off the Warning&lt;/h3&gt;
&lt;div class="outline-text-3" id="text-org0113c1a"&gt;
&lt;p&gt;
The first thing I tried was to follow the directions in the output and supress the warnings by setting an environment variable.
&lt;/p&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;&lt;span class="nb"&gt;set&lt;/span&gt;&lt;span class="w"&gt; &lt;/span&gt;--universal&lt;span class="w"&gt; &lt;/span&gt;--export&lt;span class="w"&gt; &lt;/span&gt;PYDEVD_DISABLE_FILE_VALIDATION&lt;span class="w"&gt; &lt;/span&gt;&lt;span class="m"&gt;1&lt;/span&gt;
&lt;/pre&gt;&lt;/div&gt;

&lt;p&gt;
&lt;b&gt;Note:&lt;/b&gt; This is fish-shell syntax.
&lt;/p&gt;

&lt;p&gt;
I restarted the jupyter kernel and the warnings had gone away, so this much worked.
&lt;/p&gt;
&lt;/div&gt;
&lt;/div&gt;


&lt;div id="outline-container-orgc3bd71f" class="outline-3"&gt;
&lt;h3 id="orgc3bd71f"&gt;Really Turning Off the Warning&lt;/h3&gt;
&lt;div class="outline-text-3" id="text-orgc3bd71f"&gt;
&lt;p&gt;
Setting the environment variable at the command-line changes the environment for my user, but I'm running &lt;a href="https://www.emacswiki.org/emacs/EmacsAsDaemon"&gt;emacs as a daemon&lt;/a&gt; so I needed to edit the &lt;code&gt;systemctl&lt;/code&gt; file for my emacs service. I opened up the &lt;code&gt;~/.config/systemd/user/emacs.service&lt;/code&gt; file in emacs and added the line to set the environment variable for the emacs daemon.
&lt;/p&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;&lt;span class="k"&gt;[Service]&lt;/span&gt;
&lt;span class="na"&gt;Environment&lt;/span&gt;&lt;span class="o"&gt;=&lt;/span&gt;&lt;span class="s"&gt;"PYDEVD_DISABLE_FILE_VALIDATION=1"&lt;/span&gt;
&lt;/pre&gt;&lt;/div&gt;

&lt;p&gt;
Then I restarted the service.
&lt;/p&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;systemctl&lt;span class="w"&gt; &lt;/span&gt;restart&lt;span class="w"&gt; &lt;/span&gt;--user&lt;span class="w"&gt; &lt;/span&gt;emacs
&lt;/pre&gt;&lt;/div&gt;

&lt;p&gt;
Which gave me a warning that my changes to the configuration have to be re-loaded before restarting the service.
&lt;/p&gt;

&lt;pre class="example" id="org08299bb"&gt;
Warning: The unit file, source configuration file or drop-ins of emacs.service changed on disk
. Run 'systemctl --user daemon-reload' to reload units.
&lt;/pre&gt;

&lt;p&gt;
Oops.
&lt;/p&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;systemctl&lt;span class="w"&gt; &lt;/span&gt;--user&lt;span class="w"&gt; &lt;/span&gt;daemon-reload
systemctl&lt;span class="w"&gt; &lt;/span&gt;restart&lt;span class="w"&gt; &lt;/span&gt;--user&lt;span class="w"&gt; &lt;/span&gt;emacs
&lt;/pre&gt;&lt;/div&gt;

&lt;p&gt;
This time the emacs startup messages didn't have the jupyter errors so it looked like things were fixed.
&lt;/p&gt;
&lt;/div&gt;
&lt;/div&gt;
&lt;/div&gt;

&lt;div id="outline-container-org5bf600e" class="outline-2"&gt;
&lt;h2 id="org5bf600e"&gt;Swapping a Virtual Environment For pipx&lt;/h2&gt;
&lt;div class="outline-text-2" id="text-org5bf600e"&gt;
&lt;p&gt;
Suppressing the warnings &lt;i&gt;pretty much&lt;/i&gt; solved the problem, but while I was getting this fixed I was also trying to set up a USB Windows installer using &lt;a href="https://github.com/WoeUSB/WoeUSB"&gt;WoeUSB&lt;/a&gt; and found that pipx couldn't install it because of a dependency error. Pipx is good at installing some standalone python commands but it won't install things that are just libraries and it seems to sometimes also have problems installing dependencies for the commands that it &lt;i&gt;will&lt;/i&gt; install. This has come up for me before, and the old solution was just for me to install the dependencies separately  using &lt;code&gt;pip&lt;/code&gt; before trying to install whatever it was that I was installing with pipx. Now, though, since ubuntu is trying to keep you from installing python modules globally, installing the dependencies means they either have to be available through &lt;code&gt;apt&lt;/code&gt; or you have to set up a virtual environment and install them there (when I say &lt;i&gt;have to&lt;/i&gt; I mean that since that's the way I know how to do it, that's the way I have to do it, not that there aren't other ways to do it that I just don't know about).
&lt;/p&gt;

&lt;p&gt;
Doing it this way is easy enough, since I use python virtual environments a lot anyway, but then I ran into another problem which was that once I got the virtual environment set up I found out I had to run woeUSB as root, which then bypasses the whole virtual environment setup. The solution to that was to pass the full path to the virtual environment's woeUSB launcher to &lt;code&gt;sudo&lt;/code&gt;, but it took enough time experimenting with other ways to do it before I got to that step that I decided I should minimalize how much I use pipx as much as possible - and in particular I should avoid using it with my emacs setup, since emacs will sometimss just quietly fail if there's a python-based error and it's only when things don't work that I'll realize there's a problem. So I decided to go with a dedicated virtual environment instead of installing jupyter with pipx.
&lt;/p&gt;

&lt;p&gt;
This, once again was not a big deal in hindsight, but it took enough experimenting with other options before coming to the conclusion that this was the way to go that I thought I should make a note to my future self about it. To get jupyter working with jupyter-emacs:
&lt;/p&gt;

&lt;ul class="org-ul"&gt;
&lt;li&gt;create a virtual environment (&lt;code&gt;python3 -m venv emacs-environment&lt;/code&gt;) in the &lt;code&gt;.virtualenvs&lt;/code&gt; folder&lt;/li&gt;
&lt;li&gt;activate it, then use pip to install &lt;code&gt;wheels&lt;/code&gt; and &lt;code&gt;jupyter&lt;/code&gt;&lt;/li&gt;
&lt;/ul&gt;

&lt;p&gt;
In the &lt;code&gt;/.emacs.d/init.el&lt;/code&gt; file, activate the virtual environment &lt;i&gt;before&lt;/i&gt; you load emacs-jupyter or anything else that needs python:
&lt;/p&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;&lt;span class="p"&gt;(&lt;/span&gt;&lt;span class="nf"&gt;require&lt;/span&gt;&lt;span class="w"&gt; &lt;/span&gt;&lt;span class="ss"&gt;'pyvenv&lt;/span&gt;&lt;span class="p"&gt;)&lt;/span&gt;
&lt;span class="p"&gt;(&lt;/span&gt;&lt;span class="nf"&gt;pyvenv-activate&lt;/span&gt;&lt;span class="w"&gt; &lt;/span&gt;&lt;span class="s"&gt;"~/.virtualenvs/emacs-environment"&lt;/span&gt;&lt;span class="p"&gt;)&lt;/span&gt;
&lt;/pre&gt;&lt;/div&gt;

&lt;p&gt;
Then restart emacs. So far this seems to have fixed it.
&lt;/p&gt;
&lt;/div&gt;
&lt;/div&gt;

&lt;div id="outline-container-org27b8e2a" class="outline-2"&gt;
&lt;h2 id="org27b8e2a"&gt;Other Links&lt;/h2&gt;
&lt;div class="outline-text-2" id="text-org27b8e2a"&gt;
&lt;ul class="org-ul"&gt;
&lt;li&gt;EmacsWiki: Emacs As Daemon [Internet]. [cited 2023 May 28]. Available from: &lt;a href="https://www.emacswiki.org/emacs/EmacsAsDaemon#h5o-2"&gt;https://www.emacswiki.org/emacs/EmacsAsDaemon#h5o-2&lt;/a&gt;&lt;/li&gt;
&lt;li&gt;Schäfer J. pyvenv.el, Python virtual environment support for Emacs [Internet]. 2023 [cited 2023 May 28]. Available from: &lt;a href="https://github.com/jorgenschaefer/pyvenv"&gt;https://github.com/jorgenschaefer/pyvenv&lt;/a&gt;&lt;/li&gt;
&lt;/ul&gt;
&lt;/div&gt;
&lt;/div&gt;</description><category>emacs</category><category>jupyter</category><category>python</category><category>troubleshooting</category><category>ubuntu</category><guid>https://necromuralist.github.io/posts/python-311-and-emacs-jupyter-symbols-variable-is-void-error/</guid><pubDate>Thu, 25 May 2023 23:12:27 GMT</pubDate></item><item><title>PyTorch and the Unknown CUDA Error</title><link>https://necromuralist.github.io/posts/pytorch-and-the-unknown-cuda-error/</link><dc:creator>Cloistered Monkey</dc:creator><description>&lt;div id="table-of-contents" role="doc-toc"&gt;
&lt;h2&gt;Table of Contents&lt;/h2&gt;
&lt;div id="text-table-of-contents" role="doc-toc"&gt;
&lt;ul&gt;
&lt;li&gt;&lt;a href="https://necromuralist.github.io/posts/pytorch-and-the-unknown-cuda-error/#orgb1f470a"&gt;The Problem&lt;/a&gt;&lt;/li&gt;
&lt;li&gt;&lt;a href="https://necromuralist.github.io/posts/pytorch-and-the-unknown-cuda-error/#org5897120"&gt;The Symptom&lt;/a&gt;&lt;/li&gt;
&lt;li&gt;&lt;a href="https://necromuralist.github.io/posts/pytorch-and-the-unknown-cuda-error/#org8a2ae02"&gt;The Disease and Its Cure&lt;/a&gt;&lt;/li&gt;
&lt;/ul&gt;
&lt;/div&gt;
&lt;/div&gt;
&lt;div id="outline-container-orgb1f470a" class="outline-2"&gt;
&lt;h2 id="orgb1f470a"&gt;The Problem&lt;/h2&gt;
&lt;div class="outline-text-2" id="text-orgb1f470a"&gt;
&lt;p&gt;
I recently decided to get back into using neural networks again and tried to update my docker container to get &lt;a href="https://www.fast.ai/"&gt;fastai&lt;/a&gt; up and running, but couldn't get CUDA working. After a while spent trying different configurations of CUDA, pytorch, pip, conda, and on and on I eventually found out that there's some kind of problem with using CUDA after suspending and then resuming your system (at least with linux/Ubuntu). This is a documentation of that particular problem and it's fixes (fastest but not necessarily the best answer: always shutdown or reboot the machine, don't suspend and resume).
&lt;/p&gt;
&lt;/div&gt;
&lt;/div&gt;


&lt;div id="outline-container-org5897120" class="outline-2"&gt;
&lt;h2 id="org5897120"&gt;The Symptom&lt;/h2&gt;
&lt;div class="outline-text-2" id="text-org5897120"&gt;
&lt;p&gt;
This is what happens if I try to use CUDA after waking the machine from a suspend.
&lt;/p&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;&lt;span class="kn"&gt;import&lt;/span&gt; &lt;span class="nn"&gt;torch&lt;/span&gt;

&lt;span class="n"&gt;torch&lt;/span&gt;&lt;span class="o"&gt;.&lt;/span&gt;&lt;span class="n"&gt;cuda&lt;/span&gt;&lt;span class="o"&gt;.&lt;/span&gt;&lt;span class="n"&gt;is_available&lt;/span&gt;&lt;span class="p"&gt;()&lt;/span&gt;
&lt;/pre&gt;&lt;/div&gt;

&lt;pre class="example"&gt;
/home/athena/.conda/envs/neurotic-fastai/lib/python3.9/site-packages/torch/cuda/__init__.py:88: UserWarning: CUDA initialization: CUDA unknown error - this may be due to an incorrectly set up environment, e.g. changing env variable CUDA_VISIBLE_DEVICES after program start. Setting the available devices to be zero. (Triggered internally at /opt/conda/conda-bld/pytorch_1666642975993/work/c10/cuda/CUDAFunctions.cpp:109.)
  return torch._C._cuda_getDeviceCount() &amp;gt; 0
&lt;/pre&gt;


&lt;p&gt;
As you can see, the error message doesn't really give any useful information about what's wrong - there &lt;i&gt;are&lt;/i&gt; a couple of suggestions but neither seems relevant or at least doesn't lead you to the fix.
&lt;/p&gt;
&lt;/div&gt;
&lt;/div&gt;

&lt;div id="outline-container-org8a2ae02" class="outline-2"&gt;
&lt;h2 id="org8a2ae02"&gt;The Disease and Its Cure&lt;/h2&gt;
&lt;div class="outline-text-2" id="text-org8a2ae02"&gt;
&lt;p&gt;
There's a post on the &lt;a href="https://discuss.pytorch.org/t/userwarning-cuda-initialization-cuda-unknown-error-this-may-be-due-to-an-incorrectly-set-up-environment-e-g-changing-env-variable-cuda-visible-devices-after-program-start-setting-the-available-devices-to-be-zero/129335/2"&gt;pytorch discussion boards&lt;/a&gt; about this error in which "ptrblck" says that he runs into this problem if his machine is put into the suspend state. While mentioning this he also says that restarting his machine fixes the problem, but restarting it every time seems to defeat the purpose of using suspend (and I'd have to walk to a different room to log in and decrypt the drive after restarting the machine - ugh, so much work). 
&lt;/p&gt;

&lt;p&gt;
Luckily, in a &lt;a href="https://discuss.pytorch.org/t/userwarning-cuda-initialization-cuda-unknown-error-this-may-be-due-to-an-incorrectly-set-up-environment-e-g-changing-env-variable-cuda-visible-devices-after-program-start-setting-the-available-devices-to-be-zero/129335/5"&gt;later post&lt;/a&gt; in the thread the same user mentions that you can also fix it by reloading the &lt;code&gt;nvidia_uvm&lt;/code&gt; kernel module by entering these commands in the terminal:
&lt;/p&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;sudo rmmod nvidia_uvm
sudo modprobe nvidia_uvm
&lt;/pre&gt;&lt;/div&gt;

&lt;p&gt;
Which seems to fix the problem for me right at the moment, without the need to restart the machine.
&lt;/p&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;&lt;span class="nb"&gt;print&lt;/span&gt;&lt;span class="p"&gt;(&lt;/span&gt;&lt;span class="n"&gt;torch&lt;/span&gt;&lt;span class="o"&gt;.&lt;/span&gt;&lt;span class="n"&gt;cuda&lt;/span&gt;&lt;span class="o"&gt;.&lt;/span&gt;&lt;span class="n"&gt;is_available&lt;/span&gt;&lt;span class="p"&gt;())&lt;/span&gt;
&lt;/pre&gt;&lt;/div&gt;

&lt;pre class="example"&gt;
False
&lt;/pre&gt;


&lt;p&gt;
Ummm… oops. Well, it did sort of fix one problem - the &lt;code&gt;CUDA unknown error&lt;/code&gt;, but now it's saying that CUDA isn't available on this machine. Every fix begets a new problem. Let's try it again after restarting the Jupyter kernel.
&lt;/p&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;&lt;span class="kn"&gt;import&lt;/span&gt; &lt;span class="nn"&gt;torch&lt;/span&gt;
&lt;span class="nb"&gt;print&lt;/span&gt;&lt;span class="p"&gt;(&lt;/span&gt;&lt;span class="n"&gt;torch&lt;/span&gt;&lt;span class="o"&gt;.&lt;/span&gt;&lt;span class="n"&gt;cuda&lt;/span&gt;&lt;span class="o"&gt;.&lt;/span&gt;&lt;span class="n"&gt;is_available&lt;/span&gt;&lt;span class="p"&gt;())&lt;/span&gt;
&lt;/pre&gt;&lt;/div&gt;

&lt;pre class="example"&gt;
True
&lt;/pre&gt;


&lt;p&gt;
Okay, that's better, I guess. It feels a little inelegant to have to do this, but at least it seems to work.
&lt;/p&gt;
&lt;/div&gt;
&lt;/div&gt;</description><category>cuda</category><category>pytorch</category><category>troubleshooting</category><guid>https://necromuralist.github.io/posts/pytorch-and-the-unknown-cuda-error/</guid><pubDate>Tue, 01 Nov 2022 00:58:27 GMT</pubDate></item><item><title>Mozilla Madness: Resist Fingerprinting!</title><link>https://necromuralist.github.io/posts/mozilla-madness-resist-fingerprinting/</link><dc:creator>Cloistered Monkey</dc:creator><description>&lt;div id="table-of-contents" role="doc-toc"&gt;
&lt;h2&gt;Table of Contents&lt;/h2&gt;
&lt;div id="text-table-of-contents" role="doc-toc"&gt;
&lt;ul&gt;
&lt;li&gt;&lt;a href="https://necromuralist.github.io/posts/mozilla-madness-resist-fingerprinting/#org0e173e0"&gt;The Short Version For My Future Self&lt;/a&gt;&lt;/li&gt;
&lt;li&gt;&lt;a href="https://necromuralist.github.io/posts/mozilla-madness-resist-fingerprinting/#org1a1714c"&gt;Back To the Story: It Was a Dark and Rainy Day&lt;/a&gt;&lt;/li&gt;
&lt;li&gt;&lt;a href="https://necromuralist.github.io/posts/mozilla-madness-resist-fingerprinting/#org08eb036"&gt;Start With the Nuclear Option&lt;/a&gt;&lt;/li&gt;
&lt;li&gt;&lt;a href="https://necromuralist.github.io/posts/mozilla-madness-resist-fingerprinting/#org7fc9a47"&gt;The Slow Crawl Back&lt;/a&gt;&lt;/li&gt;
&lt;li&gt;&lt;a href="https://necromuralist.github.io/posts/mozilla-madness-resist-fingerprinting/#org8432fe2"&gt;Is It a Bug?&lt;/a&gt;&lt;/li&gt;
&lt;/ul&gt;
&lt;/div&gt;
&lt;/div&gt;

&lt;div id="outline-container-org0e173e0" class="outline-2"&gt;
&lt;h2 id="org0e173e0"&gt;The Short Version For My Future Self&lt;/h2&gt;
&lt;div class="outline-text-2" id="text-org0e173e0"&gt;
&lt;p&gt;
Although some sites tell you to set Firefox's &lt;code&gt;privacy.resistFingerprinting&lt;/code&gt; option to true, it breaks altair's interaction and some other sites that use the canvas. It's probably better not to use that option, but if you do either:
&lt;/p&gt;

&lt;ul class="org-ul"&gt;
&lt;li&gt;Install the &lt;a href="https://addons.mozilla.org/en-US/firefox/addon/toggle-resist-fingerprinting/"&gt;Toggle Resist Fingerprinting&lt;/a&gt; extension and turn it off when things break (or just keep turning it off in &lt;code&gt;about:config&lt;/code&gt;).&lt;/li&gt;
&lt;li&gt;Or set &lt;code&gt;privacy.resistFingerprinting.autoDeclineNoUserInputCanvasPrompts&lt;/code&gt; to False and accept the popup requests for the page you want to use.&lt;/li&gt;
&lt;/ul&gt;
&lt;/div&gt;
&lt;/div&gt;

&lt;div id="outline-container-org1a1714c" class="outline-2"&gt;
&lt;h2 id="org1a1714c"&gt;Back To the Story: It Was a Dark and Rainy Day&lt;/h2&gt;
&lt;div class="outline-text-2" id="text-org1a1714c"&gt;
&lt;p&gt;
I decided to give &lt;a href="https://altair-viz.github.io/"&gt;altair&lt;/a&gt;, the python data visualization library a try yesterday, just to see what it looked like. I ran their "hello, world" example and managed to get a plot.
&lt;/p&gt;

&lt;object type="text/html" data="https://necromuralist.github.io/posts/mozilla-madness-resist-fingerprinting/test-bar-plot.html" style="width:100%" height="800"&gt;
  &lt;p&gt;Figure Missing&lt;/p&gt;
&lt;/object&gt;

&lt;p&gt;
Nothing fancy. If you move your cursor over the bars you might get a tool-tip giving you the width of the bar. If you do then you don't have the problem I ran into yesterday when I was trying it out. The image itself came out clear enough, but I couldn't figure out how to make the tool-tips work. I got desperate enough to try and read the documentation but it seems to be split between examples and API descriptions with little more in the way of explanatory documentation that would help to figure out how it was supposed to work. There are a lot of examples, though, so I decided to see if they would help, but then when I was looking at their &lt;a href="https://altair-viz.github.io/gallery/scatter_tooltips.html"&gt;Scatter Plot with Tool-tips&lt;/a&gt; example I noticed that their plots didn't have tool-tips either, which seemed suspicious. Was their library that broken? Was the internet?
&lt;/p&gt;

&lt;p&gt;
I took a look at the JavaScript console and that's when I saw these messages.
&lt;/p&gt;

&lt;p&gt;
&lt;img src="https://necromuralist.github.io/posts/mozilla-madness-resist-fingerprinting/blocked-canvas-data.png" alt="nil" loading="lazy"&gt;
&lt;/p&gt;

&lt;p&gt;
It looked like it might be important, but when I went searching for the message I couldn't find anything relevant. At least not at first.
&lt;/p&gt;
&lt;/div&gt;
&lt;/div&gt;

&lt;div id="outline-container-org08eb036" class="outline-2"&gt;
&lt;h2 id="org08eb036"&gt;Start With the Nuclear Option&lt;/h2&gt;
&lt;div class="outline-text-2" id="text-org08eb036"&gt;
&lt;p&gt;
My first thought was that they had somehow made altair Chromium-only so I installed &lt;a href="https://brave.com/"&gt;brave&lt;/a&gt; and, sure enough, the tool-tips worked when I switched browsers. So my initial conclusion was that I'd have to switch browsers if I decided to use altair. But then it occurred to me that I'd had problems in the past with some sites and anti-tracking options turned on in firefox so maybe it had something to do with that or one of the extensions. The question was, what setting was it or what extension? I eventually decided it was too much work to figure out so I first tried to use &lt;a href="https://support.mozilla.org/en-US/kb/diagnose-firefox-issues-using-troubleshoot-mode"&gt;Troubleshoot Mode&lt;/a&gt; to disable all the extensions, and when that didn't work, I did a &lt;a href="https://support.mozilla.org/en-US/kb/refresh-firefox-reset-add-ons-and-settings"&gt;refresh&lt;/a&gt; and wiped out all the customizations I'd done to firefox. Amazingly, this worked, but now I had to go about setting Firefox up again while avoiding whatever I did that broke altair.
&lt;/p&gt;
&lt;/div&gt;
&lt;/div&gt;

&lt;div id="outline-container-org7fc9a47" class="outline-2"&gt;
&lt;h2 id="org7fc9a47"&gt;The Slow Crawl Back&lt;/h2&gt;
&lt;div class="outline-text-2" id="text-org7fc9a47"&gt;
&lt;p&gt;
I decided to follow the advice on the &lt;a href="https://restoreprivacy.com/firefox-privacy/"&gt;restoreprivacy.com&lt;/a&gt; page, just because it came up on the first page of my search results and it seemed to touch most of the bases that I'd run before this episode. As I did a step in the setup I would check back with the altair plot to make sure that the tool-tip was still working until, eventually, I came to the setting that broke it - &lt;code&gt;privacy.resistFingerprinting&lt;/code&gt;.
&lt;/p&gt;

&lt;p&gt;
&lt;img src="https://necromuralist.github.io/posts/mozilla-madness-resist-fingerprinting/resist-fingerprinting.png" alt="nil" loading="lazy"&gt;
&lt;/p&gt;

&lt;p&gt;
When I set it to true the tool-tips would break and when I set it to false they would work again. So, then what? I didn't want to disable fingerprint protection, so I thought I'd do a little more searching and see if there was another way.
&lt;/p&gt;
&lt;/div&gt;
&lt;/div&gt;

&lt;div id="outline-container-org8432fe2" class="outline-2"&gt;
&lt;h2 id="org8432fe2"&gt;Is It a Bug?&lt;/h2&gt;
&lt;div class="outline-text-2" id="text-org8432fe2"&gt;
&lt;p&gt;
I decided to do a search on &lt;a href="https://bugzilla.mozilla.org/describecomponents.cgi?product=Firefox"&gt;Bugzilla&lt;/a&gt; and found what seemed like a relevant bug: &lt;a href="https://bugzilla.mozilla.org/show_bug.cgi?id=1631673"&gt;WhatsApp Web images broken if you flip `privacy.resistFingerprinting` due to canvas prompts without user interaction&lt;/a&gt;. The discussion is about What's App and also mentions Instagram and Twitter, and although they are focused on images, the actual error seemed close enough to what I was seeing that it seemed like it might be the same or a similar thing. But the bug was opened two years ago, so if it is a bug it doesn't seem to be something they're eager to fix. Then I ran across this bug: &lt;a href="https://bugzilla.mozilla.org/show_bug.cgi?id=1376865"&gt;Do not display Canvas Prompt unless triggered by user input&lt;/a&gt; which discusses changing the default behavior to not prompt the user for permission to use the canvas. Now that I'm describing it I'm not sure how I came to this next step based on that bug, but for some reason I went looking in &lt;code&gt;about:config&lt;/code&gt; again and noticed that right under &lt;code&gt;resistFingerprinting&lt;/code&gt; was &lt;code&gt;resistFingerprinting.autoDeclineNoUserInputCanvasPrompts&lt;/code&gt;:
&lt;/p&gt;

&lt;p&gt;
&lt;img src="https://necromuralist.github.io/posts/mozilla-madness-resist-fingerprinting/auto-decline.png" alt="nil" loading="lazy"&gt;
&lt;/p&gt;

&lt;p&gt;
This option is set to True by default and seemed to be what they were talking about in the bug so I turned it off and went back to the altair page and this time when I put my cursor over the plot a popup came up asking me for permission.
&lt;/p&gt;

&lt;p&gt;
&lt;img src="https://necromuralist.github.io/posts/mozilla-madness-resist-fingerprinting/allow-dialog.png" alt="nil" loading="lazy"&gt;
&lt;/p&gt;

&lt;p&gt;
Once allowed it, it worked, even with &lt;code&gt;resistFingerprinting&lt;/code&gt; set to true. So, if I understand what the bug reports were saying, what was causing the problem is that firefox is getting a request for the canvas but it decides that it isn't the user initiating it, so it needs extra permission but by default the box to ask for permission is turned off and the request is declined without the user (me) getting any feedback. This &lt;i&gt;seemed&lt;/i&gt; like a bug.
&lt;/p&gt;

&lt;p&gt;
As I was thinking about this I remembered that a couple of days ago I went to the kindle cloud reader and one of the books wouldn't render. I went back and played with turning &lt;code&gt;resistFingerprinting&lt;/code&gt; on and off before trying to load the ebook and this &lt;i&gt;was&lt;/i&gt; apparently the culprit so it isn't just altair that's affected for me.
&lt;/p&gt;

&lt;p&gt;
So it's a bug, right? It seemed like one, but there were already reports of this phenomena for other sites going back years so they appear to have done it on purpose. I was trying to figure out whether it was something that should be reported or not when I came upon this bug: &lt;a href="https://bugzilla.mozilla.org/show_bug.cgi?id=1666160"&gt;Users enable `privacy.resistFingerprinting` and then are surprised when it causes problems&lt;/a&gt;.
&lt;/p&gt;

&lt;p&gt;
&lt;img src="https://necromuralist.github.io/posts/mozilla-madness-resist-fingerprinting/surprise-bug.png" alt="nil" loading="lazy"&gt;
&lt;/p&gt;

&lt;p&gt;
Despite the fact that I've seen multiple sites saying to enable this "feature" maybe it isn't really a good idea after all. In the near-term I installed the &lt;a href="https://addons.mozilla.org/en-US/firefox/addon/toggle-resist-fingerprinting/"&gt;Toggle Resist Fingerprinting&lt;/a&gt; extension and use it to turn &lt;code&gt;resistFingerprinting&lt;/code&gt; on and off. To be honest, I'm not convinced that it really matters, I just got sucked into a trail of sites making conflicting assertions about what to do and convinced myself that I cared. At least I can read kindle books in the browser again.
&lt;/p&gt;
&lt;/div&gt;
&lt;/div&gt;</description><category>firefox</category><category>note-to-future-self</category><category>troubleshooting</category><category>whining</category><guid>https://necromuralist.github.io/posts/mozilla-madness-resist-fingerprinting/</guid><pubDate>Thu, 23 Dec 2021 21:54:12 GMT</pubDate></item></channel></rss>